ray-curator/ray_curator/stages/deduplication/fuzzy/connected_components.py

```diff
@@ -0,0 +1,203 @@
+# limitations under the License.
+
+import math
+from typing import TYPE_CHECKING, Any
+
+import cudf
+from cugraph.dask.comms.comms_wrapper import init_subcomms as c_init_subcomms
+from loguru import logger
+from pylibcugraph import GraphProperties, MGGraph, ResourceHandle
+from pylibcugraph import weakly_connected_components as pylibcugraph_wcc
+
+from ray_curator.backends.experimental.utils import RayStageSpecKeys
+from ray_curator.stages.base import ProcessingStage
+from ray_curator.stages.deduplication.id_generator import (
+    CURATOR_DEDUP_ID_STR,
+)
+from ray_curator.stages.deduplication.io_utils import DeduplicationIO
+from ray_curator.stages.resources import Resources
+from ray_curator.tasks.file_group import FileGroupTask
+from ray_curator.utils.file_utils import create_or_overwrite_dir, get_fs
+
+if TYPE_CHECKING:
+    from ray_curator.backends.base import WorkerMetadata
+
+
+class ConnectedComponentsStage(ProcessingStage[FileGroupTask, FileGroupTask], DeduplicationIO):
+    def __init__(
+        self,
+        output_dir: str,
+        source_column: str = f"{CURATOR_DEDUP_ID_STR}_x",
+        destination_column: str = f"{CURATOR_DEDUP_ID_STR}_y",
+        read_kwargs: dict | None = None,
+        write_kwargs: dict | None = None,
+    ):
+        """
+        Args:
+            output_dir: The directory to write the resulting connected components to.
+            source_column: The column name containing the document ids of the source of the edge.
+            destination_column: The column name containing the document ids of the destination of the edge.
+            read_kwargs: Keyword arguments to pass for reading the input files.
+            write_kwargs: Keyword arguments to pass for writing the output files.
+        """
+
+        self.source_column = source_column
+        self.destination_column = destination_column
+        self.read_kwargs = read_kwargs if read_kwargs is not None else {}
+        self.write_kwargs = write_kwargs if write_kwargs is not None else {}
+
+        self._name = self.__class__.__name__
+        self._resources = Resources(cpus=1.0, gpus=1.0)
+        self._batch_size = None
+
+        # Handle output directory cleanup logic
+        self.output_fs = get_fs(output_dir, self.write_kwargs.get("storage_options", {}))
+        self.output_dir = self.output_fs.sep.join([output_dir, self.name])
+        create_or_overwrite_dir(self.output_dir, self.output_fs)
+
+    def setup(self, _worker_metadata: "WorkerMetadata | None" = None) -> None:
+        if not hasattr(self, "_raft_handle"):
+            msg = "RAFT handle not found. Make sure the stage is initialized with RAFT"
+            raise ValueError(msg)
+
+        self._setup_post()
+
+    def ray_stage_spec(self) -> dict[str, Any]:
+        return {
+            RayStageSpecKeys.IS_RAFT_ACTOR: True,
+        }
+
+    def __get_2D_div(self, ngpus: int) -> tuple[int, int]:  # noqa: N802
```

Comment:
why `__`?

Reply:
Keeping the same convention as the source from cugraph where I adapted this from.